parser-typechecker/src/Unison/Server/CodebaseServer.hs

```diff
@@ -207,82 +191,33 @@ ucmHostVar = "UCM_HOST"
 ucmTokenVar :: String
 ucmTokenVar = "UCM_TOKEN"
 
--- The auth token required for accessing the server is passed to the function k
-start
-  :: Var v
-  => Rt.Runtime v
-  -> Codebase IO v Ann
-  -> (Strict.ByteString -> Port -> IO ())
-  -> IO ()
-start rt codebase k = do
-  envToken <- lookupEnv ucmTokenVar
-  envHost  <- lookupEnv ucmHostVar
-  envPort  <- (readMaybe =<<) <$> lookupEnv ucmPortVar
-  envUI    <- lookupEnv ucmUIVar
-  args     <- getArgs
-  let
-    p =
-      (,,,,)
-        <$> ( many $ argument str internal )
-        <*> (   (<|> envToken)
-            <$> (  optional
-                .  strOption
-                $  long "token"
-                <> metavar "STRING"
-                <> help "API auth token"
-                )
-            )
-        <*> (   (<|> envHost)
-            <$> (  optional
-                .  strOption
-                $  long "host"
-                <> metavar "STRING"
-                <> help "UCM server host"
-                )
-            )
-        <*> (   (<|> envPort)
-            <$> (  optional
-                .  option auto
-                $  long "port"
-                <> metavar "NUMBER"
-                <> help "UCM server port"
-                )
-            )
-        <*> (   (<|> envUI)
-            <$> (optional . strOption $ long "ui" <> metavar "DIR" <> help
-                  "Path to codebase ui root"
-                )
-            )
-    mayOpts =
-      getParseResult $ execParserPure defaultPrefs (info p forwardOptions) args
-  case mayOpts of
-    Just (_, token, host, port, ui) -> startServer rt codebase k token host port ui
-    Nothing -> startServer rt codebase k Nothing Nothing Nothing Nothing
+data CodebaseServerOpts = CodebaseServerOpts
+  { token :: Maybe String
+  , host :: Maybe String
+  , port :: Maybe Int
+  , codebaseUIPath :: Maybe FilePath
+  } deriving (Show)
 
+-- The auth token required for accessing the server is passed to the function k
 startServer
   :: Var v
-  => Rt.Runtime v
+  => CodebaseServerOpts
+  -> Rt.Runtime v
   -> Codebase IO v Ann
   -> (Strict.ByteString -> Port -> IO ())
-  -> Maybe String
-  -> Maybe String
-  -> Maybe Port
-  -> Maybe String
   -> IO ()
-startServer rt codebase k envToken envHost envPort envUI0 = do
+startServer opts rt codebase k = do
   -- the `canonicalizePath` resolves symlinks
   exePath <- canonicalizePath =<< getExecutablePath
-  envUI <- canonicalizePath $ fromMaybe (FilePath.takeDirectory exePath </> "ui") envUI0
-  token <- case envToken of
+  envUI <- canonicalizePath $ fromMaybe (FilePath.takeDirectory exePath </> "ui") (codebaseUIPath opts)
+  token <- case token opts of
     Just t -> return $ C8.pack t
     _      -> genToken
-  let settings = appEndo
-        (  foldMap (Endo . setPort)              envPort
-        <> foldMap (Endo . setHost . fromString) envHost
-        )
-        defaultSettings
-      a = app rt codebase envUI token
-  case envPort of
+  let settings = defaultSettings
```

Comment:
This change wasn't strictly necessary, I figure `maybe` might be more familiar to contributors than `endo` but who knows 🤷🏼 ; annoying that there isn't an idiomatic way to do this 😢